### PR TITLE
Fixing exchange bootstrap script with new GitHub and CircleCI api changes

### PR DESCRIPTION
**Pull request by nmaludy**
GitHub and CircleCI broke their APIs.

GitHub doesn't allow us to create PATs anymore via the API, so we need to do that manually. I added a new environment variable `GITHUB_PACK_APT` that is checked and a helpful error message is output.

Also, for whatever reason the CircleCI API for following repos keeps returning 400 Bad Request. I've created a support ticket with them, so we'll see where that goes. It doesn't appear to affect anything as the project shows as "followed" by the stackstorm-neptr user.

On CircleCI i switched over to using their `Circle-Token` header instead of deprecated (no longer working) URL query parameter.

### CODE REVIEW

File: utils/exchange-bootstrap.sh [hunk 1]
```diff
@@ -119,42 +135,53 @@ then
...
+# echo "CircleCI: Following the project"
+# curl -v -sS --fail -X POST \
+#      --header "Circle-Token: ${CIRCLECI_TOKEN}" \
+#      "https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/follow"
+
 
 # CircleCI: upload the read-write key
 echo "CircleCI: Adding read-write SSH key"
 curl -sS --fail -X POST --header "Content-Type: application/json" \
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:149]: inconsistent whitespace (tabs vs spaces strikes again)

File: utils/exchange-bootstrap.sh [hunk 2]
```diff
@@ -119,42 +135,53 @@ then
...
 echo "CircleCI: Setting credentials (machine login and user-scoped token)"
 curl -sS --fail -X POST --header "Content-Type: application/json" \
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
 	-d '{"name":"MACHINE_USER", "value":"'"${USERNAME}"'"}' \
-	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar?circle-token=${CIRCLECI_TOKEN}"
+	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar"
 curl -sS --fail -X POST --header "Content-Type: application/json" \
-	-d '{"name":"MACHINE_PASSWORD", "value":'"$(cat "/tmp/${PACK}_user_token")"'}' \
-	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar?circle-token=${CIRCLECI_TOKEN}"
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:160]: `s/^ */\t/`

File: utils/exchange-bootstrap.sh [hunk 3]
```diff
@@ -119,42 +135,53 @@ then
...
+	-d '{"name":"MACHINE_PASSWORD", "value":"'"$(cat "/tmp/${PACK}_user_token")"'"}' \
+	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar"
 
 # CircleCI: Enable builds for pull requests from forks
 echo "CircleCI: Enabling builds for pull requests from forks"
 curl -sS --fail -X PUT --header "Content-Type: application/json" \
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:167]: `s/^ */\t/`

File: utils/exchange-bootstrap.sh [hunk 4]
```diff
@@ -38,6 +41,23 @@ REPO_DIR="/tmp/${REPO_NAME}"
 REPO_URL="https://${USERNAME}:${PASSWORD}@github.com/${EXCHANGE_ORG}/${REPO_NAME}"
 ALIAS_URL="https://${USERNAME}:${PASSWORD}@github.com/${EXCHANGE_ORG}/${REPO_ALIAS}"
 
+# Generating GitHub PAT (personal access tokens) can no longer be done via the API
+# broken: https://developer.github.com/changes/2/#--deprecating-oauth-authorization-api
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:45]: We can probably reuse `tools/reset_github_user_token_and_update_circleci.sh` here once #107 is merged.
Something like this:

`tools/reset_github_user_token_and_update_circleci.sh --set-user ${PACK}`

File: utils/exchange-bootstrap.sh [hunk 5]
```diff
@@ -119,42 +135,53 @@ then
...
 curl -sS --fail -X POST --header "Content-Type: application/json" \
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
 	-d '{"name":"MACHINE_USER", "value":"'"${USERNAME}"'"}' \
-	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar?circle-token=${CIRCLECI_TOKEN}"
+	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar"
 curl -sS --fail -X POST --header "Content-Type: application/json" \
-	-d '{"name":"MACHINE_PASSWORD", "value":'"$(cat "/tmp/${PACK}_user_token")"'}' \
-	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/envvar?circle-token=${CIRCLECI_TOKEN}"
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
+	-d '{"name":"MACHINE_PASSWORD", "value":"'"$(cat "/tmp/${PACK}_user_token")"'"}' \
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:161]:
```suggestion
	-d '{"name":"MACHINE_PASSWORD", "value":"'"${GITHUB_PACK_PAT}"'"}' \
```

File: utils/exchange-bootstrap.sh [hunk 6]
```diff
@@ -79,21 +99,17 @@ curl -sS --fail -u "${USERNAME}:${PASSWORD}" -X POST --header "Content-Type: app
 	"https://api.github.com/repos/${EXCHANGE_ORG}/${REPO_NAME}/keys"
 
 # GitHub: create a user-scope token
-echo "Github: Creating a Github user-scoped token"
-curl -sS --fail -u "${USERNAME}:${PASSWORD}" -X POST --header "Content-Type: application/json" \
-	-d '{"scopes": ["public_repo"], "note": "CircleCI: '"${REPO_NAME}"'"}' \
-	"https://api.github.com/authorizations" | jq ".token" > "/tmp/${PACK}_user_token"
-
+echo -n "${GITHUB_PACK_PAT}" > "/tmp/${PACK}_user_token"
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:102]: Drop writing this to a file. When it's already in a var just use it from the var.

File: utils/exchange-bootstrap.sh [hunk 7]
```diff
@@ -119,42 +135,53 @@ then
...
 
 # CircleCI: upload the read-write key
 echo "CircleCI: Adding read-write SSH key"
 curl -sS --fail -X POST --header "Content-Type: application/json" \
+  --header "Circle-Token: ${CIRCLECI_TOKEN}" \
 	-d '{"hostname":"github.com","private_key":"'"$(cat "/tmp/${PACK}_rsa")"'"}' \
-	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/ssh-key?circle-token=${CIRCLECI_TOKEN}"
+	"https://circleci.com/api/v1.1/project/github/${EXCHANGE_ORG}/${REPO_NAME}/ssh-key"
 
 # CircleCI: specify the credentials (the machine login and the new user-scope token)
```
Comment (cognifloyd) [inline utils/exchange-bootstrap.sh:153]: Setting MACHINE_USER and MACHINE_PASSWORD is handled well by `tools/reset_github_user_token_and_update_circleci.sh`. We should standardize on one script so we don't duplicate these calls in both.

**Review by cognifloyd — COMMENTED** [review]
Here are a variety of thoughts. Mostly I think we need to use `tools/reset_github_user_token_and_update_circleci.sh` or extract the common logic into a separate utility shared by both tools.

File: utils/exchange-bootstrap.sh (hunk 1 above)
Comment (armab) [reply utils/exchange-bootstrap.sh:149]: We're using 2 whitespaces across the st2 fleet for bash/script files and it indeed would be great following that in all the other places.

Mixing both (like this script) is indeed confusing.

**Review by cognifloyd — COMMENTED** [review]
A couple of dangling references to `GITHUB_PACK_PAT` need to be removed.

**Review by cognifloyd — APPROVED** [review]

**Pull request merged**